common/include/camera/ICameraControl.hpp

```diff
@@ -0,0 +1,312 @@
+        // Orientation of gimbal
+        inline const glm::quat& getOrientation() const { return m_orientation; }
+
+        // Orthonormal [getXAxis(), getYAxis(), getZAxis()] matrix
+        inline const float32_t3x3& getOrthonornalMatrix() const { return m_orthonormal; }
+
+        // Base right vector in orthonormal basis, base "right" vector (X-axis)
+        inline const float32_t3& getXAxis() const { return m_orthonormal[0u]; }
+
+        // Base up vector in orthonormal basis, base "up" vector (Y-axis)
+        inline const float32_t3& getYAxis() const { return m_orthonormal[1u]; }
+
+        // Base forward vector in orthonormal basis, base "forward" vector (Z-axis)
+        inline const float32_t3& getZAxis() const { return m_orthonormal[2u]; }
+
+    private:
+        inline void updateOrthonormalMatrix() 
+        { 
+            m_orthonormal = float32_t3x3(glm::mat3_cast(glm::normalize(m_orientation))); 
+
+            // DEBUG
+            const auto [xaxis, yaxis, zaxis] = std::make_tuple(getXAxis(),getYAxis(), getZAxis());
+
+            auto isNormalized = [](const auto& v, float epsilon) -> bool
+            {
+                return glm::epsilonEqual(glm::length(v), 1.0f, epsilon);
+            };
+
+            auto isOrthogonal = [](const auto& a, const auto& b, float epsilon) -> bool
+            {
+                return glm::epsilonEqual(glm::dot(a, b), 0.0f, epsilon);
+            };
+
+            auto isOrthoBase = [&](const auto& x, const auto& y, const auto& z, float epsilon = 1e-6f) -> bool
+            {
+                return isNormalized(x, epsilon) && isNormalized(y, epsilon) && isNormalized(z, epsilon) &&
+                    isOrthogonal(x, y, epsilon) && isOrthogonal(x, z, epsilon) && isOrthogonal(y, z, epsilon);
+            };
+
+            assert(isOrthoBase(xaxis, yaxis, zaxis));
+        }
+
+        float32_t3 m_position;
+        glm::quat m_orientation;
+
+        // Represents the camera's orthonormal basis
+        // https://en.wikipedia.org/wiki/Orthonormal_basis
+        float32_t3x3 m_orthonormal;
+    };
+
+    ICameraController(core::smart_refctd_ptr<CGimbal>&& gimbal) : m_gimbal(core::smart_refctd_ptr(gimbal)) {}
+
+    // override controller keys map, it binds a key code to a virtual event
+    void updateKeysToEvent(const std::vector<ui::E_KEY_CODE>& codes, VirtualEventType event)
+    {
+        m_keysToEvent[event] = std::move(codes);
+    }
+
+    // start controller manipulation session
+    virtual void begin(std::chrono::microseconds nextPresentationTimeStamp)
+    {
+        m_nextPresentationTimeStamp = nextPresentationTimeStamp;
+        return;
+    }
+
+    // manipulate camera with gimbal & virtual events, begin must be called before that!
+    virtual void manipulate(std::span<const CVirtualEvent> virtualEvents) = 0;
+
+    // finish controller manipulation session, call after last manipulate in the hot loop
+    void end(std::chrono::microseconds nextPresentationTimeStamp)
```

Comment:
the begin/end don't actually affect `manipulate` because the events already have a magnitude/impulse by them.

its actually event  generation/processing/translation that youre beginning/ending that need and updates the timestamps

Reply:
resolved